job/server/src/main/java/alluxio/job/plan/transform/format/CachedPath.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.job.plan.transform.format;
+
+import alluxio.client.ReadType;
+import alluxio.client.WriteType;
+import alluxio.conf.InstancedConfiguration;
+import alluxio.conf.PropertyKey;
+
+import com.google.common.base.Objects;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Implementation of {@link Path} that has a cache for getting file system.
+ */
+public class CachedPath extends Path {
+
+  private static final ConcurrentHashMap<FileSystemKey, FileSystem> CACHE =
+      new ConcurrentHashMap<>();
+
+  private static class FileSystemKey {
+
+    final String mScheme;
+    final String mAuthority;
+    final UserGroupInformation mUgi;
+    final ReadType mReadType;
```

Comment:
Are these the only user config we are interested in? Could there be issues for reusing a client when other user config values differ?

Reply:
There definitely can be new user configs in the future that we may care that we currently don't. It's going to be impossible to have this be always up-to-date. For example, one that I can currently think of Replication.